### ISSUE TITLE
gradient noise vs variational weight noise

### ISSUE THREAD

**Issue by qhoangdl** [NVIDIA/NeMo#3831]
Hi, I notice that the variational noise for RNNT model is implemented as gradient noise addition (line 878 of `nemo/collections/asr/models/rnnt_models.py`). The variational noise cited by the Conformer-Transducer paper is actually adding noise to weights. So, is this a mistake or because injecting gradient noise works better than adding weight noise?

**Comment by titu1994** [comment]
Weight + noise didn't show much effect on training in our runs, whereas gradient + noise did sometimes show better regularization (though too much and it would hurt results like all over regularizes models)

**Comment by qhoangdl** [comment]
Hi @titu1994 , thanks for the answer. I'd like to experiment with variational weight noise but I'm struggling to find an elegant way to implement as I need to change every layer from h = wx to h = (w + noise)x. Simply performing in-place addition of noises to the weights won't work because we want the optimizer to update the weight before adding noise rather than the weight after adding noise. So, do you have any suggestion?

**Comment by titu1994** [comment]
This is how Lingvo implements it - and as far as I can see there is no easy way to do so without modifying the source code of every single block that uses this - https://github.com/tensorflow/lingvo/blob/master/lingvo/core/py_utils.py#L3484-L3553

You also don't modify w itself, but the input x becomes x + noise * scale before passing into the learnable parameter.

**Comment by qhoangdl** [comment]
Thanks @titu1994 . I'm thinking of something like this:

```
<At the begining of each training step>
noise = {}
with torch.no_grad():
    for param in model.parameters():
        noise[param.name] = torch.randn(param.size()) * variational_noise_std
        param.add_(noise[param.name])
...
<at the end of each training step, right after the weight update>
with torch.no_grad():
    for param in model.parameters():
        noise[param.name] = torch.randn(param.size()) * variational_noise_std
        param.sub_(noise[param.name])
```
This way, we can still compute the gradient for the sampled weight (i.e. weight + noise) while making update to the weights before adding noise. But I'm not sure if this is an efficient/elegant way of implementing it.

**Comment by VahidooX** [comment]
Actually I tried adding the noise to the weights but results was worse than gradient noise. Here is the experimental branch I tested this technique: 

https://github.com/VahidooX/NeMo/tree/add_weight_noise

In the paper, there was not much detail on variational noise. I even tried just adding that to the decoder part, still got no improvement. Not sure if it is a good idea to add noise to all the weights. I gussed that they just used it for the RNN or the decoder. Here is how I did it:

https://github.com/VahidooX/NeMo/blob/c3f4ac8e4585af2fb94faad13aa05775b3057f24/nemo/collections/asr/models/rnnt_models.py#L843

You may change the following line to add the noise to the whole model or whole decoder instead of just RNN part of the decoder:

`for param_name, param in self.decoder.prediction.dec_rnn.named_parameters():`

Something like self.named_parameters() for the whole model.

**Comment by VahidooX** [comment]
BTW, I assumed that on multi-GPU training each GPU has a copy of params and changing that would not affect the master copy. Therefore decided not to revert them back. Now I am not sure if my assumption was correct specially if tested on single GPU. Maybe that is the reason it did not worked. What you can do is to store the noise in a dictionary something like self.var_noises and then use it to revert back the weights after each step. The other way is to implement a version of all layers which support variational noise like what you suggested at the beginning.

**Comment by qhoangdl** [comment]
Thanks @VahidooX! I'll try it out to see if it works.
